File: 2023/day_5/star_2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <sstream>
#include <climits>

struct SeedRange
{
    unsigned long start;
    unsigned long length;
};

struct Converter
{
    unsigned long source;
    unsigned long destination;
    unsigned long length;
};

std::vector<Converter> extractValues(const std::vector<std::string>& data, int& i) {
    std::vector<Converter> converter;
    int n = data.size();
    while (i < n && data[i].rfind(':') > data[i].size()) {
        std::stringstream ss;
        std::string src, dest, len;
        ss << data[i];
        ss >> dest;
        ss >> src;
        ss >> len;
        converter.push_back({stoul(src), stoul(dest), stoul(len)});
        i++;
    }
    i++;
    return converter;
}

unsigned long convertValue(const std::vector<Converter>& converter, unsigned long value) {
    for (Converter con : converter) {
        if (value >= con.source && value < con.source + con.length) {
            return value - con.source + con.destination;
        }
    }
    return value;
}

int main () {
    std::ifstream file("day_5/input.txt");
    std::string line;
    std::vector<std::string> data;

    while (std::getline(file, line)) {
        if (line == "\n" || line.empty()) continue;
        data.push_back(line);
    }

    std::vector<SeedRange> seeds;
    std::string start;
    std::string length;
    std::stringstream ss;
    ss << data[0];
    ss >> start;
    while (!ss.eof()) {
        ss >> start;
        ss >> length;
        SeedRange range = {stoul(start),stoul(length)};
        
        seeds.push_back(range);
    }

    int i = 2;
    std::vector<Converter> seedToSoil = extractValues(data, i);
    std::vector<Converter> soilToFertilizer = extractValues(data, i);
    std::vector<Converter> fertilizerToWater = extractValues(data, i);
    std::vector<Converter> waterToLight = extractValues(data, i);
    std::vector<Converter> lightToTemperature = extractValues(data, i);
    std::vector<Converter> temperatureToHumidity = extractValues(data, i);
    std::vector<Converter> humidityToLocation = extractValues(data, i);

    unsigned long minLocation = ULONG_MAX;
    for (SeedRange range : seeds) {
        std::cout << "Range from " << range.start << " to " << range.length << '\n';
        unsigned long value;
        for (unsigned long seed = range.start; seed < range.start + range.length; seed++) {
            value = convertValue(seedToSoil, seed);
            value = convertValue(soilToFertilizer, value);
            value = convertValue(fertilizerToWater, value);
            value = convertValue(waterToLight, value);
            value = convertValue(lightToTemperature, value);
            value = convertValue(temperatureToHumidity, value);
            value = convertValue(humidityToLocation, value);
            if (value < minLocation) minLocation = value;
        }
    }
    std::cout << minLocation << '\n';
    return 0;
}

File: 2023/day_5/star_2.py

def extract_values(data: 'list[str]', i: int) -> ('list[dict]', int):
    result = []
    while i < len(data) and not data[i].endswith(':'):
        values = data[i].split()
        result.append({'dest': int(values[0]), 'source': int(values[1]), 'length': int(values[2])})
        i += 1
    return result, i

def convert_value(converter: 'list[dict]', value: int):
    for line in converter:
        if value >= line['source'] and value < line['source'] + line['length']:
            return value - line['source'] + line['dest']
    return value

with open("day_5/input.txt", 'r') as f:
    data = f.readlines()

data = [line.strip() for line in data if line != '\n']

sum = 0
seed_line = [int(val) for val in data[0].split(':')[1].split()]
seeds = [(seed_line[i], seed_line[i+1]) for i in range(0,len(seed_line),2)]

i = 2
seed_to_soil, i = extract_values(data, i)
soil_to_fertilizer, i = extract_values(data, i + 1)
fertilizer_to_water, i = extract_values(data, i + 1)
water_to_light, i = extract_values(data, i + 1)
light_to_temp, i = extract_values(data, i + 1)
temp_to_humidity, i = extract_values(data, i + 1)
humidity_to_location, i = extract_values(data, i + 1)

min_loc = float('inf')
for seed_range in seeds:
    start = seed_range[0]
    stop = start + seed_range[1]
    for seed in range(start, stop):
        print(seed)
        soil = convert_value(seed_to_soil, seed)
        fertilizer = convert_value(soil_to_fertilizer, soil)
        water = convert_value(fertilizer_to_water, fertilizer)
        light = convert_value(water_to_light, water)
        temperature = convert_value(light_to_temp, light)
        humidity = convert_value(temp_to_humidity, temperature)
        location = convert_value(humidity_to_location, humidity)
        if location < min_loc:
            min_loc = location

print(min_loc)